### PR TITLE
Feature/hashtags

### LINKED ISSUE #76: As a user, I want to use hashtags in posts, so that they are in that topic
#65 is blocking this feature

### PR DESCRIPTION
**Pull request by juanecabellob**
hashtag feature done; it is almost possible to highlight mentions as well, but first is first.

Closes #76.

### CODE REVIEW

**Comment by juanecabellob** [comment]
just realized that i forgot the documentation of the code; i will fix that tomorrow :smile:

**Comment by metalmatze** [comment]
I don't know if this was the case yesterday. But you should also merge the current develop into your branch, so that I don't have to merge all the conflicts, since you know your code base. :wink:

**Comment by juanecabellob** [comment]
I had this done actually, yesterday said actually "this merge can be done automatically" ..

**Comment by juanecabellob** [comment]
As you can see in the commits, I merged the develop branch.

File: posts/templates/posts/posts_list.html [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+                        <p class="text-muted lead" style="color: #aaa; text-align: center;">
+                            This user has no posts<br/>
+                            #circuitzombie</p>
+                    {% endif %}
+                </div>
+            </div>
+
+        </div>
+    </div>
+{% endblock %}
```
Comment (Gjum) [inline posts/templates/posts/posts_list.html]: Why did you copy the profile page for this?

Also missing newline at end of file.

File: posts/templatetags/__init__.py [hunk 2]
```diff
@@ -0,0 +1 @@
+__author__ = 'juancabello'
```
Comment (Gjum) [inline posts/templatetags/__init__.py]: This should be in a separate PR, as we did not finally decide about the format of docuentation.

File: posts/templatetags/parser.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+from django.core.urlresolvers import reverse
+from posts.views import PostsListView
+from django.template.defaultfilters import stringfilter
+import re
+
+register = template.Library()
+
+@register.filter
+@stringfilter
+def parse(content):
```
Comment (Gjum) [inline posts/templatetags/parser.py]: The name of this filter is too ambiguous. `parse` could mean anything.
Something like `parse_post_content` or `insert_hashtag_links` would make it clearer what this filter does.

File: posts/urls.py [hunk 4]
```diff
@@ -8,4 +9,5 @@
     url(r'^(?P<pk>\d+)/edit/$', views.PostEditView.as_view(), name='edit'),
     url(r'^(?P<pk>\d+)/repost/$', views.PostRepostView, name='repost'),
     url(r'^(?P<pk>\d+)/favorite/$', views.PostFavoriteView, name='favorite'),
+    url(r'^show/(?P<hashtag_name>\w+)/$', PostsListView.as_view(), name='posts_list'),
```
Comment (Gjum) [inline posts/urls.py]: What does `show` mean? Does it only show hashtags? Please rename.

File: posts/views.py [hunk 5]
```diff
@@ -86,6 +91,39 @@ def PostRepostView(request, pk=None):
     return redirect('posts:post', pk=repost.pk)
 
 
+def ParseContent(content):
```
Comment (Gjum) [inline posts/views.py:94]: Damn @MetalMatze, why did you even start this? :D Now everyone names their functions in UpperCamelCase! We should keep it this way for now and mass-rename them all later to `snake_case`.

File: posts/views.py [hunk 6]
```diff
@@ -86,6 +91,39 @@ def PostRepostView(request, pk=None):
...
+    hashtags = re.findall(r"#(\w+)", content)
+    mentions = re.findall(r"@(\w+)", content)
```
Comment (Gjum) [inline posts/views.py]: This is the hashtag feature, please try next time to not include unrelated code in a PR.

File: posts/templatetags/__init__.py (hunk 2 above)
Comment (metalmatze) [inline posts/templatetags/__init__.py]: I think this file should be completely empty. I think PyCharm created that file.

File: users/models.py [hunk 7]
```diff
@@ -1,18 +1,39 @@
+"""@package docstring
+User model.
+
+@author     7Pros
+@copyright  Some license
+"""
```
Comment (Gjum) [inline users/models.py:6]: This should be in a separate PR, as it is not related to the hashtag feature in any way.

Same for the other documentation in this file.

File: posts/templatetags/parser.py (hunk 3 above)
Comment (metalmatze) [inline posts/templatetags/parser.py]: Yep, good point! :+1:

File: posts/urls.py (hunk 4 above)
Comment (metalmatze) [inline posts/urls.py]: This would correspond to a url like `localhost:8000/posts/show/foobar`?
I'd prefer something like `localhost:8000/posts/hashtag/foobar` even better would be `localhost:8000/posts?hashtag=foobar` or `localhost:8000/hashtag/foobar`.
The last one would not be in this `posts/urls.py` but in `circuit/urls.py`.

Which one do you prefer?

File: posts/views.py (hunk 5 above)
Comment (metalmatze) [inline posts/views.py:94]: You're right. Sorry. Will do refactoring tomorrow after merging all pull requests for this sprint.

File: posts/models.py [hunk 8]
```diff
@@ -17,3 +17,18 @@ def original_or_self(self):
...
             p = p.original_post
         return p
+
+    def __str__(self):
+        return self.content
+
+class Hashtag(models.Model):
+    name = models.CharField(max_length=255)
+    posts = models.ManyToManyField(Post)
+    created_at = models.DateTimeField(auto_now_add=True)
```
Comment (metalmatze) [inline posts/models.py]: Seems like we don't need this `created_at` timestamp for a hashtag, since `Post` already has the same timestamps.

Comment (Gjum) [inline posts/models.py]: Wooo, more migrations! :D

File: posts/urls.py (hunk 4 above)
Comment (Gjum) [inline posts/urls.py]: `/posts/hashtags/foobar` (note the **s**) would be my personal choice, as it's the closest to REST.
My reasoning is that you look for posts (not e.g. users), and in the posts, you look for hashtags (not e.g. mentions or authors).
`/hashtags/foobar` would also be ok, but hashtag search should be encapsulated into the posts app IMO.

File: posts/templatetags/parser.py [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+    pattern = re.compile(r"#(\w+).*?", flags=re.IGNORECASE | re.UNICODE)
+    parsed_content = pattern.sub(createHashtagURL, content)
+
+    return parsed_content
```
Comment (Gjum) [inline posts/templatetags/parser.py:20]: This filter does not work for special chars that get HTML-escaped, at least `&'"<>` and probably [even more](https://en.wikipedia.org/wiki/List_of_XML_and_HTML_character_entity_references#Character_entity_references_in_HTML).

When posting `I'm here` for example, the result gets rendered as `I&#39;m here`, with the `#39` highlighted as a hashtag.

The corresponding HTML looks like this:

``` html
I&amp;
<a class="hashtag" href="/posts/show/39/">#39</a>
;m here
```

This is seems to be the result of searching the escaped text (`I&#39;m here`, seems to have the hashtag `#39`) for hashtags instead of the actually typed text (`I'm here`, has no hashtags).

**Comment by Gjum** [comment]
TODO before merge:
- [ ] fix hashtag filter for special chars
- [ ] rename hashtag filter
- [ ] change hashtag URL to `/posts/hashtags/<hashtag name>`
- [ ] delete the content of `posts/templatetags/__init__.py`

File: posts/models.py (hunk 8 above)
Comment (metalmatze) [inline posts/models.py]: Maybe @juanecabellob could delete the current ones for this feature and simply push the changes in the migration. I don't think it's necessary to have a new migration, since it's not merged yet.

File: posts/templates/posts/posts_list.html (hunk 1 above)
Comment (juanecabellob) [inline posts/templates/posts/posts_list.html]: I copied it, because I needed to check if the filter was working and I thought twitter makes it just as I did - just realized it doesn't, they have a completely new view for this. What we could do, is just take the profile part away, leaving the posts alone.

File: posts/models.py (hunk 8 above)
Comment (juanecabellob) [inline posts/models.py]: It will create a new migration, even if I do it myself.

File: posts/urls.py (hunk 4 above)
Comment (juanecabellob) [inline posts/urls.py]: I agree with this variant: /posts/hashtags/<hashtag_name>
